published/external/xdpapi.h

```diff
@@ -264,9 +264,72 @@ XDP_RSS_GET_FN(
     _Inout_ UINT32 *RssConfigurationSize
     );
 
-typedef struct _XDP_QEO_CONFIGURATION {
+typedef enum _XDP_QUIC_OPERATION {
+    XDP_QUIC_OPERATION_ADD,     // Add (or modify) a QUIC connection offload
+    XDP_QUIC_OPERATION_REMOVE,  // Remove a QUIC connection offload
+} XDP_QUIC_OPERATION;
+
+typedef enum _XDP_QUIC_DIRECTION {
+    XDP_QUIC_DIRECTION_TRANSMIT, // An offload for the transmit path
+    XDP_QUIC_DIRECTION_RECEIVE,  // An offload for the receive path
+} XDP_QUIC_DIRECTION;
+
+typedef enum _XDP_QUIC_DECRYPT_FAILURE_ACTION {
+    XDP_QUIC_DECRYPT_FAILURE_ACTION_DROP,     // Drop the packet on decryption failure
+    XDP_QUIC_DECRYPT_FAILURE_ACTION_CONTINUE, // Continue and pass the packet up on decryption failure
+} XDP_QUIC_DECRYPT_FAILURE_ACTION;
+
+typedef enum _XDP_QUIC_CIPHER_TYPE {
+    XDP_QUIC_CIPHER_TYPE_AEAD_AES_128_GCM,
+    XDP_QUIC_CIPHER_TYPE_AEAD_AES_256_GCM,
+    XDP_QUIC_CIPHER_TYPE_AEAD_CHACHA20_POLY1305,
+    XDP_QUIC_CIPHER_TYPE_AEAD_AES_128_CCM,
+} XDP_QUIC_CIPHER_TYPE;
+
+typedef enum _XDP_QUIC_ADDRESS_FAMILY {
+    XDP_QUIC_ADDRESS_FAMILY_INET4,
+    XDP_QUIC_ADDRESS_FAMILY_INET6,
+} XDP_QUIC_ADDRESS_FAMILY;
+
+typedef struct _XDP_QUIC_CONNECTION {
     XDP_OBJECT_HEADER Header;
-} XDP_QEO_CONFIGURATION;
+    UINT32 Operation            : 1;  // XDP_QUIC_OPERATION
+    UINT32 Direction            : 1;  // XDP_QUIC_DIRECTION
+    UINT32 DecryptFailureAction : 1;  // XDP_QUIC_DECRYPT_FAILURE_ACTION
+    UINT32 KeyPhase             : 1;
+    UINT32 RESERVED             : 12; // Must be set to 0. Don't read.
+    UINT32 CipherType           : 16; // XDP_QUIC_CIPHER_TYPE
+    XDP_QUIC_ADDRESS_FAMILY AddressFamily;
+    UINT16 UdpPort;         // Destination port.
+    UINT64 NextPacketNumber;
+    UINT8 ConnectionIdLength;
+    UINT8 Address[16];      // Destination IP address.
+    UINT8 ConnectionId[20]; // QUIC v1 and v2 max CID size
+    UINT8 PayloadKey[32];   // Length determined by CipherType
+    UINT8 HeaderKey[32];    // Length determined by CipherType
+    UINT8 PayloadIv[12];
+    HRESULT Status;         // The result of trying to offload this connection.
+} XDP_QUIC_CONNECTION;
+
+#define XDP_QUIC_CONNECTION_REVISION_1 1
+
+#define XDP_SIZEOF_QUIC_CONNECTION_REVISION_1 \
+    RTL_SIZEOF_THROUGH_FIELD(XDP_QUIC_CONNECTION, Status)
+
+//
+// Initializes a QEO configuration object.
+//
+inline
+VOID
+XdpInitializeQuicConnection(
+    _Out_writes_bytes_(XdpQuicConnectionSize) XDP_QUIC_CONNECTION *XdpQuicConnection,
+    _In_ UINT32 XdpQuicConnectionSize
+    )
+{
+    RtlZeroMemory(XdpQuicConnection, XdpQuicConnectionSize);
+    XdpQuicConnection->Header.Revision = XDP_QUIC_CONNECTION_REVISION_1;
+    XdpQuicConnection->Header.Size = XDP_SIZEOF_QUIC_CONNECTION_REVISION_1;
```

Comment:
Instead of per-connection version, how about a struct that wraps the array of connections, and the struct has the verison? We will always push down at least 2 connections at a time, possibly more; so it will be more efficient to have one version apply to all of them.

Reply:
Yep. However, I'm thinking this approach may minimize the amount of code rewriting that needs to occur in a future model where each QUIC connection gets pushed onto an IO ring for the user/kernel transition. It is easier to deal with fixed-size blocks (i.e., single connections) being pushed onto a ring than variable-sized arrays.

Reply:
Basically, we can think about this API as an interim solution, where batching is performed by sending an array of elements across down to the XDP driver. As we add more offloads and focus more on the control path performance, we may instead push and pull single elements onto a shared ring, and then process the ring in batches, similar to packets.

Reply:
It's unclear how this proposed IO ring style approach would work and why it would result in fixed-sized blocks. I assume this IO ring would be for multiple different types of OIDs, and not just 1 ring per OID type? Currently, the caller assumes this code is blocking, because we cannot progress with the connection until we know if the call succeeds or fails. By moving to an IO ring model, I assume it becomes async and we have to deal with that on the QUIC side too?

Reply:
Yeah, my assumption is that this API will be async in the future, as soon as the locking bugs in the XDP offload module are fixed. Async APIs can be trivially make synchronous if the caller is fine with blocking, but the converse is not true. Note that this OID is not required to be processed especially quickly by the hardware, so we want to avoid the need to have dedicate worker threads idly waiting for blocking IOs to be completed.

Re: IO ring, yeah, I think we'd want to have a common control path / IOCTL ring. It's not that a versioned struct containing an array is incompatible with that approach, I'm just contending that the efficiency gains by versioning a struct containing two elements vs versioning each element is unlikely to be measurable. I'd rather think of each connection independently, which also simplifies how we'll need to convert and track these in the XDP kernel.

Reply:
Ok. Thanks for the explanation.